File: basestation/Backend/DEFCOM/ChannelLossyCast.py
import threading
from ._FlexibleMessageStructure import MessageStructure
from ._DefComParser import ConnectionSpecification as _ConnectionSpecification, loadConfFile as _loadConfFile
from ._MultiUDPWrapper import udpsend, udpget

import time
import os
import struct

class LossyCastPublisher:
    #Initialise with the message structure definition (defcom file) and a function pointer
    #The function must be of the type:  void function(MessageStructure request, MessageStructure response)
    def __init__(self, defComFile: str):
        self._definition: _ConnectionSpecification = _loadConfFile(defComFile)

        if self._definition.RequestMessageFormat.totalSize > 1000:
            raise Exception('Message size too large for UDP')
        
        self._con = udpsend('239.0.0.1', self._definition.NumericPort, self._definition.RequestMessageFormat.totalSize + 8)

        self.sequence = 63

    #Get a message to be filled in
    def getNewMessageObject(self) -> MessageStructure:
        return self._definition.RequestMessageFormat.clone()

    def publish(self,requestData: MessageStructure):
        #Get the buffer
        buffer = requestData.getDecodeBuffer()
        buffer = struct.pack('Q',self.sequence) + buffer

        self._con.senddat(buffer)

        #Add a sequence header to it
        self.sequence += 1
        


    


class LossyCastSubscriber:
    def __init__(self, defComFile: str):
        self._definition: _ConnectionSpecification = _loadConfFile(defComFile)

        if self._definition.RequestMessageFormat.totalSize > 1000:
            raise Exception('Message size too large for UDP')

        #Connect to the server
        self._con = udpget('239.0.0.1', self._definition.NumericPort, self._definition.RequestMessageFormat.totalSize + 8)

        self.lastSequence = 0
        

    def subscribe(self) -> MessageStructure:
        #Receive the response
        while True:
            message = self._con.getdat()

            #Get the sequence
            sequence = struct.unpack('Q',message[:8])[0]

            # Sequencing is scary, thankfully some engineers smarter than me worked this out
            # Out sequence ring is 64 with 32 being the division between old and new
            # This means that it can take max 32 packets for the connection to recover after loss
            if ((sequence - self.lastSequence) % 64) < 32: 
                self.lastSequence = sequence
                message = message[8:]
                break
            
        #Return the response
        response = self._definition.RequestMessageFormat.clone()
        
        if message:
            response.setDecodeBuffer(message)

        return response

    

File: onboard/defnodes/DriveInterface/external/DEFCOM/src/ChannelLossycast.cpp
#include "ChannelLossycast.hpp"
#include "comms/Multicast.hpp"
#include "DefComParser.hpp"
#include "FlexibleMessageStructure.hpp"
#include <string>
#include <stdint.h>

LossyCastPublisher::LossyCastPublisher(std::string filePath){
    this->definition = loadConfFile(filePath);

    if (this->definition.RequestMessageFormat.totalSize > 1000){
        throw std::runtime_error("Message size too large for UDP");
    }

    sock = new udpsend("239.0.0.1", this->definition.NumericPort, this->definition.RequestMessageFormat.totalSize + 8);
    sequence = 63;
}

LossyCastPublisher::~LossyCastPublisher(){
    delete sock;
}

MessageStructure LossyCastPublisher::getNewMessageObject(){
    return MessageStructure(this->definition.RequestMessageFormat);
}

void LossyCastPublisher::publish(MessageStructure requestData){
    unsigned char outgoing[this->definition.RequestMessageFormat.totalSize + 8];
    uint64_t snum = sequence;
    memcpy(outgoing, &snum, 8);
    memcpy(outgoing + 8, requestData.getDecodeBuffer(), this->definition.RequestMessageFormat.totalSize);
    
    sequence++;
    sock->senddat(outgoing);
}


LossyCastSubscriber::LossyCastSubscriber(std::string filePath){
    this->definition = loadConfFile(filePath);

    if (this->definition.RequestMessageFormat.totalSize > 1000){
        throw std::runtime_error("Message size too large for UDP");
    }

    sock = new udpget("239.0.0.1", this->definition.NumericPort, this->definition.RequestMessageFormat.totalSize + 8);
    lastSequence = 0;
}

LossyCastSubscriber::~LossyCastSubscriber(){
    delete sock;
}

MessageStructure LossyCastSubscriber::subscribe(){
    unsigned char* incoming;
    while (true){
        incoming = sock->getdat();

        uint64_t sequence;
        memcpy(&sequence, incoming, 8);

        if (((sequence - this->lastSequence) % 64) < 32) {
            lastSequence = sequence;
            break;
        }

    }

    MessageStructure output = MessageStructure(this->definition.RequestMessageFormat);
    output.setDecodeBuffer(incoming + 8);
    
    return output;


}